File: iso-lab/e2e.mjs
import { JSDOM } from 'jsdom';
import { readFileSync } from 'node:fs';

const html = readFileSync(new URL('./index.html', import.meta.url), 'utf8');
// the page was SSR'd for the deep link /shop/products/7 (app mounted at base /shop)
const dom = new JSDOM(html, {
  url: 'http://localhost/shop/products/7',
  runScripts: 'dangerously',
  pretendToBeVisual: true,
  beforeParse(w) { w.TextDecoder = globalThis.TextDecoder; w.TextEncoder = globalThis.TextEncoder; },
});
const { document } = dom.window;

let pass = 0, fail = 0;
const eq = (label, got, want) => {
  if (String(got) === String(want)) { pass++; console.log(`  ✓ ${label} = ${got}`); }
  else { fail++; console.log(`  ✗ ${label}: got ${JSON.stringify(got)}, want ${JSON.stringify(want)}`); }
};
const wait = ms => new Promise(r => setTimeout(r, ms));

const page       = () => document.querySelector('.page')?.getAttribute('data-page');
const h1         = () => document.querySelector('.page h1')?.textContent;
const headTitle  = () => document.querySelector('head title')?.textContent;
const titleCount = () => document.querySelectorAll('head title').length;
const descCount  = () => document.querySelectorAll('head meta[name="description"]').length;
const og         = () => document.querySelector('head meta[property="og:title"]')?.getAttribute('content');
const attr       = (sel, n) => document.querySelector(sel)?.getAttribute(n);
const text       = sel => document.querySelector(sel)?.textContent;
const path       = () => dom.window.location.pathname;
const fire       = el => el.dispatchEvent(new dom.window.MouseEvent('click', { bubbles: true, cancelable: true }));
const click      = sel => fire(document.querySelector(sel));
const counterBtn = i => document.querySelector('.counter').querySelectorAll('button')[i]; // 0=−, 1=+
const todos      = () => document.querySelectorAll('.todos li').length;
const stats      = () => text('.stats');
// jump to an arbitrary in-app URL the way the browser would (history + popstate)
const goto = p => { dom.window.history.pushState({}, '', p); dom.window.dispatchEvent(new dom.window.Event('popstate')); };

(async () => {
  await wait(120);

  console.log('— deep-link hydration: product page, base-relative route —');
  eq('page', page(), 'product');
  eq('param', h1(), 'Product #7');
  eq('title', headTitle(), 'Product 7 · shop');
  eq('location', path(), '/shop/products/7');
  eq('one <title>', titleCount(), 1);
  eq('layout og survives across pages', og(), 'iso shop');

  console.log('— nav hrefs: typed p (base-prefixed) + ext (outer reach) —');
  eq('nav-home', attr('.nav-home', 'href'), '/shop');
  eq('nav-products', attr('.nav-products', 'href'), '/shop/products');
  eq('nav-admin (ext, no base)', attr('.nav-admin', 'href'), '/admin');
  eq('back link', attr('.back', 'href'), '/shop/products');

  console.log('— back link → products —');
  click('.back'); await wait(20);
  eq('page', page(), 'products');
  eq('title', headTitle(), 'Products · shop');
  eq('location', path(), '/shop/products');
  eq('one <title> (Head cleanup on unmount)', titleCount(), 1);
  eq('one description', descCount(), 1);
  eq('product links built by p', attr('.p7', 'href'), '/shop/products/7');
  eq('product links built by p', attr('.p9', 'href'), '/shop/products/9');

  console.log('— FORM: type into the input (reads target_value), add, controlled clear —');
  eq('stats initial', stats(), 'todos in store: 0');
  const typeAdd = txt => {
    const input = document.querySelector('#todo-input');
    input.value = txt;
    input.dispatchEvent(new dom.window.Event('input', { bubbles: true }));  // -> onInput reads target_value
    click('#add');
  };
  typeAdd('milk'); typeAdd('eggs');
  eq('two typed todos', todos(), 2);
  eq('typed value used (not a fixed string)', document.querySelector('.todos li').firstChild.textContent.trim(), 'milk');
  eq('controlled input cleared after add', document.querySelector('#todo-input').value, '');
  eq('stats', stats(), 'todos in store: 2');
  click('.todos li .rm');                 // remove first (keyed reconcile)
  eq('one todo after remove', todos(), 1);
  eq('now first is eggs', document.querySelector('.todos li').firstChild.textContent.trim(), 'eggs');
  eq('stats after remove', stats(), 'todos in store: 1');

  console.log('— reactive param: click Product 9 —');
  click('.p9'); await wait(20);
  eq('page', page(), 'product');
  eq('param', h1(), 'Product #9');
  eq('title tracks param', headTitle(), 'Product 9 · shop');
  eq('location', path(), '/shop/products/9');
  eq('still one <title> after 3 page visits', titleCount(), 1);

  console.log('— Home: local counter (+/−), client data, client-only, on_mount —');
  click('.nav-home'); await wait(120);
  eq('page', page(), 'home');
  eq('title', headTitle(), 'Home · shop');
  fire(counterBtn(1)); fire(counterBtn(1)); fire(counterBtn(0));   // + + −
  eq('counter local state (1+1−1=1)', text('.counter .count'), 1);
  eq('greeting fetched on client nav', text('.greeting .msg'), 'Hello again, from the client 🔁');
  eq('greeting ready', text('.greeting .gstatus').trim(), '(ready)');
  eq('client-only data', text('.browser-box .bdata'), 'Loaded in the browser ✨');
  eq('on_mount ran', text('.browser-box .mounted'), 'mounted ✓');
  eq('localStorage visit counter (Browser facade)', text('.visits'), 'visits (localStorage): 1');
  eq('localStorage actually written', dom.window.localStorage.getItem('visits'), '1');

  console.log('— greeting refetch: real dynamic fetch (loading → ready) —');
  click('#refetch');
  eq('refetch shows loading', text('.greeting .gstatus').trim(), '(loading)');
  await wait(60);
  eq('refetch resolved', text('.greeting .msg'), 'Hello again, from the client 🔁');
  eq('refetch ready again', text('.greeting .gstatus').trim(), '(ready)');

  console.log('— catch-all: unknown in-app path → not found page —');
  goto('/shop/nope/here'); await wait(20);
  eq('page', page(), 'notfound');
  eq('catch-all param', text('.missing'), 'no route: /nope/here');  // path keeps its leading /
  eq('location', path(), '/shop/nope/here');

  console.log('— global store persisted across all that navigation (no reload) —');
  click('.nav-products'); await wait(20);
  eq('page', page(), 'products');
  eq('todo survived', todos(), 1);
  eq('stats', stats(), 'todos in store: 1');

  console.log('— popstate (browser back) —');
  dom.window.history.back(); await wait(30);   // back to /shop/nope/here
  eq('popstate navigates', page(), 'notfound');
  eq('popstate location', path(), '/shop/nope/here');

  console.log('— outer-reach (ext) link is NOT hijacked —');
  click('.nav-admin'); await wait(20);
  eq('ext left our route untouched', page(), 'notfound');

  console.log('— SSR payload empty for the product deep-link —');
  eq('empty data context', /window\.__ISO_DATA__=\{\}/.test(html), true);

  console.log('— SECOND APP (/admin) reuses the shared <Counter/> —');
  const adminHtml = readFileSync(new URL('./admin.html', import.meta.url), 'utf8');
  const adm = new JSDOM(adminHtml, {
    url: 'http://localhost/admin', runScripts: 'dangerously', pretendToBeVisual: true,
    beforeParse(w) { w.TextDecoder = globalThis.TextDecoder; w.TextEncoder = globalThis.TextEncoder; },
  });
  await wait(120);
  const ad = adm.window.document;
  eq('admin page rendered', ad.querySelector('.page')?.getAttribute('data-page'), 'admin-home');
  eq('admin title', ad.querySelector('head title')?.textContent, 'Admin · dashboard');
  eq('admin reuses shared Counter', !!ad.querySelector('.counter'), true);
  // the SAME component is live in the admin app — increment it
  ad.querySelector('.counter').querySelectorAll('button')[1]
    .dispatchEvent(new adm.window.MouseEvent('click', { bubbles: true }));
  eq('shared Counter works in admin', ad.querySelector('.counter .count').textContent, 1);
  eq('admin location is its own base', adm.window.location.pathname, '/admin');

  console.log(`\n${fail === 0 ? 'PASS' : 'FAIL'} — ${pass} passed, ${fail} failed`);
  process.exit(fail === 0 ? 0 : 1);
})();

File: examples/site/e2e/e2e.ml
(* Real-browser end-to-end test for the site example — pure OCaml/Eio, ZERO npm. It
   boots the native server.exe, launches a headless Chrome, and drives it over the Chrome
   DevTools Protocol (see cdp.ml — a hand-written WebSocket+CDP client over Eio) to prove
   the things SSR alone can't: actual js_of_ocaml HYDRATION, local + global signal state,
   isomorphic data + fast-render + refetch, the Browser/localStorage facade, on_mount,
   controlled form inputs, SPA navigation, dynamic params, the catch-all, and strict
   per-app bundle isolation.

   Run with examples/site/e2e/run.sh (builds deps, then execs this). Override the browser
   with CHROME=… and the server binary with argv(1). *)

let chrome_default = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
let cdp_port = 9456
let web = "http://localhost:8200"
let admin = "http://localhost:8201"

let fails = ref 0
let check name ok = if ok then Printf.printf "  ok   %s\n%!" name else (incr fails; Printf.printf "  FAIL %s\n%!" name)

let starts p s = String.length s >= String.length p && String.sub s 0 (String.length p) = p

(* tiny HTTP/1.1 GET over Eio. The server is keep-alive, so we MUST honour Content-Length
   (reading to EOF would block) — read the status line + headers, then exactly N bytes. *)
let http_get ~sw net ~port ~path =
  let flow = Eio.Net.connect ~sw net (`Tcp (Eio.Net.Ipaddr.V4.loopback, port)) in
  Eio.Flow.copy_string
    (Printf.sprintf "GET %s HTTP/1.1\r\nHost: localhost:%d\r\nConnection: close\r\n\r\n" path port)
    (flow :> _ Eio.Flow.sink);
  let r = Eio.Buf_read.of_flow (flow :> _ Eio.Flow.source) ~max_size:(8 * 1024 * 1024) in
  let _status = Eio.Buf_read.line r in
  let clen = ref None in
  let rec headers () =
    match Eio.Buf_read.line r with
    | "" -> ()
    | h ->
      let lower = String.lowercase_ascii h in
      if starts "content-length:" lower then
        clen := int_of_string_opt (String.trim (String.sub h 15 (String.length h - 15)));
      headers ()
  in
  headers ();
  match !clen with Some n -> Eio.Buf_read.take n r | None -> Eio.Buf_read.take_all r
let ws_path url =
  let u = if starts "ws://" url then String.sub url 5 (String.length url - 5) else url in
  match String.index_opt u '/' with Some i -> String.sub u i (String.length u - i) | None -> "/"

(* retry a thunk until it returns without raising, or time out *)
let wait_ready clock ~desc ~timeout f =
  let deadline = Eio.Time.now clock +. timeout in
  let rec loop () =
    match f () with
    | x -> x
    | exception e -> if Eio.Time.now clock > deadline then failwith (desc ^ ": " ^ Printexc.to_string e)
                     else (Eio.Time.sleep clock 0.1; loop ())
  in
  loop ()

let () =
  Eio_main.run @@ fun env ->
  Eio.Switch.run @@ fun sw ->
  let net = Eio.Stdenv.net env and clock = Eio.Stdenv.clock env in
  let pmgr = Eio.Stdenv.process_mgr env and fs = Eio.Stdenv.fs env in
  let server_exe = if Array.length Sys.argv > 1 then Sys.argv.(1) else "_build/default/examples/site/server.exe" in
  let chrome = try Sys.getenv "CHROME" with Not_found -> chrome_default in
  let devnull = Eio.Path.open_out ~sw ~create:(`If_missing 0o644) (Eio.Path.(/) fs "/tmp/fennec_e2e.log") in
  let spawn args = Eio.Process.spawn ~sw pmgr ~stdout:devnull ~stderr:devnull args in

  (* 1. boot the server + a headless Chrome with a CDP debug port. Start from a clean
     browser profile each run so a previously-cached bundle can never be served. *)
  ignore (Sys.command "rm -rf /tmp/fennec_e2e_chrome");
  let server = spawn [ server_exe ] in
  let chrome_proc =
    spawn
      [ chrome; "--headless=new"; "--disable-gpu"; "--no-sandbox"; "--no-first-run";
        "--user-data-dir=/tmp/fennec_e2e_chrome"; "--disk-cache-size=1"; "--disable-application-cache";
        Printf.sprintf "--remote-debugging-port=%d" cdp_port; "--remote-allow-origins=*"; "about:blank" ]
  in
  let cleanup () =
    (try Eio.Process.signal chrome_proc Sys.sigkill with _ -> ());
    (try Eio.Process.signal server Sys.sigkill with _ -> ())
  in
  Fun.protect ~finally:cleanup @@ fun () ->
  (* wait for the server to answer *)
  ignore (wait_ready clock ~desc:"server boot" ~timeout:15.0 (fun () ->
              let b = http_get ~sw net ~port:8200 ~path:"/" in
              if Cdp.contains b "Fennec" then b else failwith "not ready yet"));
  (* wait for Chrome's debug endpoint, get the browser websocket *)
  let ws_url =
    wait_ready clock ~desc:"chrome boot" ~timeout:15.0 (fun () ->
        let v = http_get ~sw net ~port:cdp_port ~path:"/json/version" in
        match Cdp.field "webSocketDebuggerUrl" (Yojson.Safe.from_string v) with
        | Some (`String u) -> u
        | _ -> failwith "no ws url")
  in
  let browser = Cdp.attach (Cdp.ws_connect ~sw net ~port:cdp_port ~path:(ws_path ws_url)) clock in
  let sess = Cdp.new_page browser in
  let t = browser in
  let pollc = Cdp.poll_contains t sess in
  let eval_s = Cdp.eval_str t sess and eval_b = Cdp.eval_bool t sess in

  Printf.printf "\n-- hydration + isomorphic data --\n%!";
  Cdp.navigate t sess (web ^ "/");
  ignore (pollc ~desc:"hydration (on_mount)" ~want:"mounted ✓" ~timeout:15.0
            "document.querySelector('.mounted').textContent");
  check "hydration: on_mount ran" true;
  check "data: fast-render value present" (Cdp.contains (eval_s "document.querySelector('.greeting .msg').textContent") "Hello from the server");
  check "data: seeded resource is ready (no flash)" (Cdp.contains (eval_s "document.querySelector('.greeting .gstatus').textContent") "(ready)");
  ignore (pollc ~desc:"client-only data fetch" ~want:"fetched live in the browser" ~timeout:5.0
            "document.querySelector('.bdata').textContent");
  check "data: client-only resource fetched after hydration" true;

  Printf.printf "-- local signal state (counter) --\n%!";
  ignore (pollc ~desc:"counter click → 1" ~want:"1" ~timeout:5.0
            "(function(){document.querySelector('.cbtn.inc').click();return document.querySelector('.count').textContent})()");
  check "counter: increments on click" true;
  ignore (pollc ~desc:"counter click → 2" ~want:"2" ~timeout:5.0
            "(function(){document.querySelector('.cbtn.inc').click();return document.querySelector('.count').textContent})()");
  ignore (pollc ~desc:"counter click → 1" ~want:"1" ~timeout:5.0
            "(function(){document.querySelector('.cbtn.dec').click();return document.querySelector('.count').textContent})()");
  check "counter: decrements on click" true;

  Printf.printf "-- Browser facade (localStorage) --\n%!";
  check "localStorage: first visit = 1" (Cdp.contains (eval_s "document.querySelector('.visits').textContent") "visits (localStorage): 1");
  Cdp.navigate t sess (web ^ "/");
  ignore (pollc ~desc:"re-hydrate" ~want:"mounted ✓" ~timeout:15.0 "document.querySelector('.mounted').textContent");
  ignore (pollc ~desc:"localStorage persists/increments" ~want:"visits (localStorage): 2" ~timeout:5.0
            "document.querySelector('.visits').textContent");
  check "localStorage: persists + increments to 2" true;

  Printf.printf "-- data refetch --\n%!";
  ignore (eval_s "document.querySelector('#refetch').click()");
  ignore (pollc ~desc:"refetch returns" ~want:"(ready)" ~timeout:5.0 "document.querySelector('.greeting .gstatus').textContent");
  check "data: refetch keeps server value" (Cdp.contains (eval_s "document.querySelector('.greeting .msg').textContent") "Hello from the server");

  Printf.printf "-- SPA navigation (no reload) --\n%!";
  ignore (eval_s "window.__spa = 1");
  ignore (eval_s "document.querySelector('.nav-link[href=\"/products\"]').click()");
  ignore (pollc ~desc:"SPA nav to /products" ~want:"/products" ~timeout:5.0 "location.pathname");
  check "spa: url changed to /products" (eval_s "location.pathname" = "/products");
  check "spa: products content rendered" (Cdp.contains (eval_s "document.body.textContent") "todos in store");
  check "spa: no full reload (window marker survived)" (eval_b "window.__spa === 1");

  Printf.printf "-- global store + forms (each/keyed) --\n%!";
  check "store: starts empty" (Cdp.contains (eval_s "document.querySelector('.stats').textContent") "todos in store: 0");
  ignore (eval_s "(function(){var e=document.querySelector('#todo-input');e.value='milk';e.dispatchEvent(new Event('input',{bubbles:true}));e.dispatchEvent(new KeyboardEvent('keydown',{key:'Enter',bubbles:true}))})()");
  ignore (pollc ~desc:"add via Enter" ~want:"todos in store: 1" ~timeout:5.0 "document.querySelector('.stats').textContent");
  check "forms: Enter adds todo, store updates Stats" true;
  check "forms+each: keyed row rendered" (Cdp.contains (eval_s "document.querySelector('.todo-items').textContent") "milk");
  ignore (eval_s "(function(){var e=document.querySelector('#todo-input');e.value='eggs';e.dispatchEvent(new Event('input',{bubbles:true}));document.querySelector('#add').click()})()");
  ignore (pollc ~desc:"add via button" ~want:"todos in store: 2" ~timeout:5.0 "document.querySelector('.stats').textContent");
  check "forms: add-button adds second todo" true;
  ignore (eval_s "document.querySelector('.todo .rm').click()");
  ignore (pollc ~desc:"remove row" ~want:"todos in store: 1" ~timeout:5.0 "document.querySelector('.stats').textContent");
  check "store: remove updates Stats" true;

  Printf.printf "-- dynamic route param (typed Paths, SPA) --\n%!";
  ignore (eval_s "document.querySelector('.p7').click()");
  ignore (pollc ~desc:"SPA nav to /products/7" ~want:"/products/7" ~timeout:5.0 "location.pathname");
  check "param: /products/7 renders Product #7" (Cdp.contains (eval_s "document.body.textContent") "Product #7");

  Printf.printf "-- catch-all (not_found) --\n%!";
  Cdp.navigate t sess (web ^ "/nope/xyz");
  ignore (pollc ~desc:"catch-all" ~want:"no route: /nope/xyz" ~timeout:10.0 "document.body.textContent");
  check "catch-all: unmatched path renders not_found with param" true;

  Printf.printf "-- strict per-app bundle isolation --\n%!";
  Cdp.navigate t sess (web ^ "/");
  ignore (pollc ~desc:"home reload" ~want:"mounted ✓" ~timeout:15.0 "document.querySelector('.mounted').textContent");
  check "isolation: web bundle EXCLUDES admin-only code"
    (not (eval_b "(async()=>(await (await fetch('/_apps/web/main.js')).text()).includes('admin actions'))()"));

  Printf.printf "-- admin app (separate endpoint, shared components, recolored) --\n%!";
  Cdp.navigate t sess (admin ^ "/");
  ignore (pollc ~desc:"admin counter present" ~want:"admin actions" ~timeout:15.0 "document.body.textContent");
  check "admin: whitelabel body class" (Cdp.contains (eval_s "document.body.className") "admin");
  ignore (pollc ~desc:"admin counter hydrates" ~want:"1" ~timeout:5.0
            "(function(){document.querySelector('.cbtn.inc').click();return document.querySelector('.count').textContent})()");
  check "admin: shared Counter hydrates independently" true;
  check "isolation: admin bundle INCLUDES admin-only code"
    (eval_b "(async()=>(await (await fetch('/_apps/admin/main.js')).text()).includes('admin actions'))()");

  Printf.printf "\n%s\n%!" (if !fails = 0 then "site e2e OK (real browser)" else Printf.sprintf "%d check(s) failed" !fails);
  if !fails <> 0 then exit 1
